internal/api/auth/ldap.go

```diff
@@ -1,41 +1,212 @@
 package auth
 
 import (
+	"context"
 	"crypto/tls"
+	"errors"
 	"fmt"
+	"time"
 
 	"github.com/rs/zerolog/log"
 
 	"github.com/go-ldap/ldap/v3"
 )
 
+type connectionStatus bool
+
+var ErrLDAPTimeout = errors.New("LDAP timeout")
+
+const (
+	connectionClosed connectionStatus = false
+	connectionUp     connectionStatus = true
+)
+
+type authRequest struct {
+	authResp chan bool
+	username string
+	password string
+}
+
+type result struct {
+	auth bool
+	conn connectionStatus
+}
+
 type LDAPAuth struct {
-	ldapClient *ldap.Conn
-	bindDN     string
-	password   string
-	baseDN     string
+	tlsConfig             *tls.Config
+	reqCh                 chan authRequest
+	ldapURL               string
+	bindDN                string
+	password              string
+	baseDN                string
+	maxConnectionLifetime time.Duration
 }
 
 func NewLDAPAuth(ldapURL string, bindDN string, password string, baseDN string, tlsConfig *tls.Config) *LDAPAuth {
-	conn, err := ldap.DialURL(ldapURL, ldap.DialWithTLSConfig(tlsConfig))
-	if err != nil {
-		log.Error().Err(err).Str("ldapURL", ldapURL).Msg("failed to connect to the LDAP server")
-		return nil
-	}
 	return &LDAPAuth{
-		ldapClient: conn,
-		bindDN:     bindDN,
-		password:   password,
-		baseDN:     baseDN,
+		tlsConfig: tlsConfig,
+		ldapURL:   ldapURL,
+		bindDN:    bindDN,
+		password:  password,
+		baseDN:    baseDN,
+		reqCh:     make(chan authRequest),
 	}
 }
 
+func SetLDAPDefaultTimeout(timeout time.Duration) {
+	ldap.DefaultTimeout = timeout //nolint:reassign  // we want to customize the default timeout
+}
+
+// SetMaxConnectionLifetime sets the maximum lifetime of a connection.
+//
+// The maximum lifetime is the maximum amount of time a connection may be reused for.
+// This is not a guarantee, as the connection may have been closed by the server before reaching that timer.
+func (l *LDAPAuth) SetMaxConnectionLifetime(maxConnectionLifetime time.Duration) {
+	l.maxConnectionLifetime = maxConnectionLifetime
+}
+
+// StartAuthenticationWorkers starts a pool of workers that will handle the authentication requests.
+func (l *LDAPAuth) StartAuthenticationWorkers(ctx context.Context, workersCount int) error {
+	if workersCount <= 0 {
+		return fmt.Errorf("'WorkersCount' must be greater than 0: %d", workersCount)
+	}
+	for i := 0; i < workersCount; i++ {
+		go l.spawnConnectionWorker(ctx)
+	}
+	return nil
+}
 func (l *LDAPAuth) AuthenticateUser(username string, password string) bool {
-	if err := l.ldapClient.Bind(l.bindDN, l.password); err != nil {
+	req := authRequest{
+		username: username,
+		password: password,
+		authResp: make(chan bool),
+	}
+	l.reqCh <- req
+	return <-req.authResp
+}
+
+func (l *LDAPAuth) spawnConnectionWorker(ctx context.Context) {
+	const maxRetry = 1
+	var conn *ldap.Conn
+	var err error
+	tick := time.NewTicker(l.maxConnectionLifetime)
+
+	for {
+		select {
+		case req := <-l.reqCh:
+			auth := false
+			retry := 0
+			for retry <= maxRetry+1 {
```

Comment:
why maxRetry +1 ? with that you'll have up to 3 attempts

Reply:
it is now clearer that 3 attempts is the wanted number (it was unclear due to iteration during tests)